pkg/kv/kvserver/replica_write.go

```diff
@@ -884,3 +889,132 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
 	ms.Reset()
 	mvccStatsPool.Put(ms)
 }
+
+// overlapsUnreplicatedRangeIDLocalKeys checks if the provided span overlaps
+// with any unreplicated rangeID local keys.
+func overlapsUnreplicatedRangeIDLocalKeys(span roachpb.Span) error {
+	// getRangeIDLocalSpan is a helper function that returns the full
+	// unreplicated RangeID span for a given RangeID.
+	getRangeIDLocalSpan := func(rangeID roachpb.RangeID) roachpb.Span {
+		return roachpb.Span{Key: keys.MakeRangeIDUnreplicatedPrefix(rangeID),
+			EndKey: keys.MakeRangeIDUnreplicatedPrefix(rangeID).PrefixEnd()}
+	}
+
+	// (1) If we receive the following span:
+	// NIL------ SPAN ---------), we convert it into:
+	//                        [), and we check that it doesn't overlap any
+	// rangeID local keys.
+	if span.Key == nil {
+		// Since it's impossible to get the Prev key to EndKey, we perform this
+		// conservative approach of making the start key a little before end key.
+		modifiedSpan := roachpb.Span{Key: span.EndKey.
+			Prevish(roachpb.PrevishKeyLength), EndKey: span.EndKey}
+
+		startRangeID, _, _, _, err := keys.DecodeRangeIDKey(modifiedSpan.Key)
+		if err != nil {
+			// If the conservative start key is not a RangeID key, then this span
+			// cannot overlap with any rangeID local keys.
+			return nil //nolint:returnerrcheck
+		}
+
+		if modifiedSpan.Overlaps(getRangeIDLocalSpan(startRangeID)) {
+			return errors.Errorf("overlapping an unreplicated rangeID span")
+		}
```

Comment:
It seems like `DecodeRangeIDKey` gives us all we need. The `infix` bit is equal to [localRangeIDUnreplicatedInfix](https://github.com/cockroachdb/cockroach/blob/2b759bf2e7d059bdb53caf40a626ad00fa8bc85e/pkg/keys/constants.go#L111) when we're in the unreplicated space. But it's a bit annoying that, due to the `(x-eps,x]` semantics of our span, we should be checking `start < x <= end` here, which means `"u" < infix+suffix+detail <= "v"`.

But we know that `infix+suffix+detail` can't equal `u` or `v`, because `suffix` must't be empty (`DecodeRangeIDKey` would return an error).

So... looks like it's sufficient to check that `infix == "u"`?